libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -509,6 +509,32 @@ bool DeclarationRegistrationHelper::registerDeclaration(
 	// We use "invisible" for both inactive variables in blocks and for members invisible in contracts.
 	// They cannot both be true at the same time.
 	solAssert(!(_inactive && !_declaration.isVisibleInContract()), "");
+
+	static set<string> illegalNames{"_", "super", "this"};
+
+	if (illegalNames.count(name))
+	{
+		auto isPublicFunctionOrEvent = [](Declaration const* _d) -> bool
+		{
+			if (auto functionDefinition = dynamic_cast<FunctionDefinition const*>(_d))
+				if (!functionDefinition->isFree() && functionDefinition->isPublic())
+					return true;
+
+			if (dynamic_cast<EventDefinition const*>(_d))
+				return true;
+
+			return false;
```

Comment:
```suggestion
			else
				return false;
```

Reply:
Fixed the other ones. The problem with this is that when the function is free or internal, this `else` branch won't be visited.

Reply:
Ah right. What about something like

```
			if (auto functionDefinition = dynamic_cast<FunctionDefinition const*>(_d); !functionDefinition->isFree() && functionDefinition->isPublic())
					return true;
			else if (dynamic_cast<EventDefinition const*>(_d))
				return true;
			else
				return false;
```

Reply:
I feel that the first if statement is less readable.